waku-wacky-whisper-spec.md

```diff
@@ -0,0 +1,46 @@
+# Waku Wacky Whisper Specification
+
+> Version: 0.1 (Draft)
+
+> Authors: Infura "The Devil" Osaka <noreply@status.im>
+
+Basically Infura for chat. Addresses scalability issues of Whisper by sacrificing metadata protection. Specifically for resource restricted devices like mobilephones with limited data plan.
+
+インフラ (Infura) means infrastructure in Japanese, and 枠 (Waku) is a closely related word that means 'frame' or 'slide'.
+
+## Rationale
+
+Many users put a premium on bandwidth usage over metadata protection. Implementing Waku mode gives users a more performant experience, at the expense of metadata protection and requiring stronger connectivity.
+
+For more background, please see Whisper theoretical model [report](https://htmlpreview.github.io/?https://github.com/vacp2p/research/blob/master/whisper_scalability/report.html) with associated [source code](https://github.com/vacp2p/research/tree/master/whisper_scalability).
+
+## Roles and definitions
+
+1. Waku-chan
+2. Waku-san (aka Waku node)
+
+A client is said to implement waku mode if it connects to a Waku node and acts as a Waku-chan. The Waku node connects to the rest of the Whisper network.
+
+## Amendments to Whisper
+
+Waku mode is compatible with [EIP-627](https://eips.ethereum.org/EIPS/eip-627) and simply extends its capabilities. It does this through a new packet code, as well as some client specific recommendations. Two specific protocol changes are done:
+
+1. Modify EIP627 by adding a new packet code, e.g. `Topic List [101, bytes]` where bytes is a list of N topics.
+
+Format TBD.
+
+2. Provide way to identify oneself as a Waku node.
```

Comment:
Is this where we need to start getting into transport negotiation or can a waku client talk to non waku using clients?

Reply:
Maybe it could be as simple as adding a field to the existing Whisper handshake, to identify as Waku-chan or Waku-san.
Behaviour at handshake would then be:
- in case of Waku-chan -> disconnect unless talking to a Waku-san. Only make 1 connection.
- in case of Waku-san:
  - when connecting with Waku-chan: apply Topic list filter when sending messages to this peer
  - when connecting with others: current behaviour (full messages or bloomfilter)

Problem with this might be how to do it without losing backwards compatibility (non waku nodes). But maybe it will just work as is, to be tested. I believe EIP-8 requests that you can add extra fields to the handshake, but that is specifically for rlpx/devp2p handshake. I don't think it mentions this for subprotocols such as Whisper.
And in fact, I think in our nim implementation that would be something we have to correct.
I'm not sure if geth its behaviour, something to test.

Additionally, I think the topic list could then also be added in the Whisper handshake (and the new packet code can be used to update it). This way we don't get a bunch of unwanted messages send right after the handshake.